### PR TITLE
[PROD-993] - Support for passphrase protected key file

### PR DESCRIPTION
**Pull request by spkesan**

### CODE REVIEW

File: aeroprom.conf.dev [hunk 1]
```diff
@@ -28,17 +29,28 @@ db_port=3000
...
 cert_file=""
+
 # key file
 key_file=""
+
+# Passphrase for encrypted key_file. Supports below formats,
+# 1. Passphrase directly                 - "<passphrase>"
+# 2. Passphrase via file                 - "file:<file-that-contains-passphrase>"
+# 3. Passphrase via environment variable - "env:<environment-variable-that-holds-passphrase>"
+key_file_passphrase="${AS_KEY_FILE_PASSPHRASE}"
```
Comment (khaf) [inline aeroprom.conf.dev:40]: Not sure this should ever be set in the config file; just from either the ENV or as a command line parameter.

File: aeroprom.conf.dev [hunk 2]
```diff
@@ -28,17 +29,28 @@ db_port=3000
...
 root_ca=""
 
 # authentication mode: internal (for server), external (LDAP, etc.)
 auth_mode=""
+
 # database user
 user=""
+
 # database password
 password=""
```
Comment (khaf) [inline aeroprom.conf.dev:55]: We may want to remove these two arguments as well. It was a mistake on my part that I didn't remove them before release.

File: observer.go [hunk 3]
```diff
@@ -57,10 +58,56 @@ func initTLS() *tls.Config {
 
 	var clientPool []tls.Certificate
 	if len(config.Aerospike.CertFile) > 0 || len(config.Aerospike.KeyFile) > 0 {
-		// client Cert
-		cert, err := tls.LoadX509KeyPair(config.Aerospike.CertFile, config.Aerospike.KeyFile)
+
```
Comment (khaf) [inline observer.go:61]: This function has become too long. I think it's not a bad idea to move the reading of the file out to a separate function.

File: aeroprom.conf.dev (hunk 1 above)
Comment (spkesan) [reply aeroprom.conf.dev:40]: Resolving as discussed. We follow the same convention in server and other tools.

File: aeroprom.conf.dev (hunk 2 above)
Comment (spkesan) [reply aeroprom.conf.dev:55]: Resolving as discussed.

File: observer.go (hunk 3 above)
Comment (spkesan) [reply observer.go:57]: Done. moved all certificate file reads into a function `readCertFile()`

File: observer.go [hunk 4]
```diff
@@ -46,21 +46,56 @@ func initTLS() *tls.Config {
...
+		// Read Cert and Key files
+		certFileBytes := readCertFile(config.Aerospike.CertFile)
+		keyFileBytes := readCertFile(config.Aerospike.KeyFile)
+
+		// Decode PEM data
+		keyBlock, _ := pem.Decode([]byte(keyFileBytes))
```
Comment (khaf) [inline observer.go:63]: Shouldn't you check errors here?

File: observer.go [hunk 5]
```diff
@@ -46,21 +46,56 @@ func initTLS() *tls.Config {
...
+
+		// Check and Decrypt the the Key Block using passphrase
+		if x509.IsEncryptedPEMBlock(keyBlock) {
+			keyFilePassphraseBytes, err := getKeyFilePassphrase(config.Aerospike.KeyFilePassphrase)
+			if err != nil {
+				log.Fatalf("Failed to get Key file passphrase for `%s` : `%s`", config.Aerospike.KeyFile, err)
+			}
+
+			decryptedDERBytes, err := x509.DecryptPEMBlock(keyBlock, keyFilePassphraseBytes)
+
```
Comment (khaf) [inline observer.go:78]: This empty line should be removed.

File: observer.go (hunk 4 above)
Comment (spkesan) [reply observer.go:63]: There's a check right after this code to see if `keyBlock` is `nil`. `keyBlock` will be `nil` if it's unable to find or decode the PEM data in the input `keyFileBytes`.
https://golang.org/pkg/encoding/pem/#Decode

File: observer.go (hunk 5 above)
Comment (spkesan) [reply observer.go:78]: Will remove the empty line.

**Pull request merged**